src/typing/env_js.ml

```diff
@@ -293,49 +342,93 @@ let already_bound_error =
    (since multiple declarations - sometimes but not always erroneous -
    may appear in an AST)
  *)
-let bind_entry cx name entry =
-  let scope = peek_scope () in
-  match Scope.get_entry name scope with
 
+let bind_entry cx name entry reason =
+  (* lex scopes can only hold let/const bindings
+   * var scopes can hold all bindings
+   * var bindings can't shadow lex bindings
+   * type entries are hoisted to var scope *)
+  let rec find_scope = function
+    | [] -> assert_false "empty scope list"
+    | scope::scopes -> Scope.(
+      match scope.kind, entry with
+      | LexScope, Entry.Value { Entry.kind = Entry.Let _; _ }
+      | LexScope, Entry.Value { Entry.kind = Entry.Const; _ }
+      | VarScope _, _ ->
+        Some scope
+      | LexScope, Entry.Type _ ->
+        find_scope scopes
+      | LexScope, Entry.Value { Entry.kind = Entry.Var; _ } ->
+        match get_entry name scope with
+        (* nested scope var shadows a lex-scoped lexical *)
+        | Some prev
+          when is_lex scope && Entry.is_lex prev ->
+            already_bound_error cx name prev reason;
+            None
+        (* keep searching for the var's scope *)
+        | _ -> find_scope scopes
```

Comment:
So, I added this "check as we walk the lex scopes" logic a bit hastily, and I should have refactored here. The reason the `LexScope, Entry.Value _` case even exists is to ensure that a var doesn't shadow a lex binding in an intermediate lex scope, see [the associated test case](https://github.com/facebook/flow/pull/784/files#diff-d5b8aa5064f80d1a4462d05b7f7fa872R173), for an example. This is also why I added the `is_lex` guard, above; we'll catch the non-`is_lex` case below.

I think the solution here is to actually merge the `find_scope` logic with the rest of the logic below, and eliminate the intermediate `Scope.t option` stuff. I should be able to add enough comments to make the intention clear.